### PR TITLE
Add unit tests for interpolation and recovery

### LINKED ISSUE #89: Fix gaps in code coverage for `recovery.py`
The improvements to code coverage brought by #86 also revealed some gaps, especially for recovery. It would be good to address these.

### PR DESCRIPTION
**Pull request by jwallwork23**
This PR reworks the interpolation and recovery routines used in Pyroteus and adds unit tests.

Partially addresses #77.

Closes #89.

### CODE REVIEW

**Comment by jwallwork23** [comment]
Tests were failing because:
* C kernel par_loops have been removed from Firedrake, so I removed them here.
* The `is_loopy_kernel=True` kwarg is no longer needed, so I dropped it.
* However, the CI is using a slightly older version of Firedrake, where this kwarg is still required.
* I have been struggling to update the Docker image used in the CI (see [139_dockerfiles](https://github.com/pyroteus/pyroteus/tree/139_dockerfiles)).

Solution for now is just to keep the `is_loopy_kernel=True`. Their removal is added as a To Do in #139.

**Comment by jwallwork23** [comment]
@stephankramer Apologies that this also became a decent sized PR. It would be good to merge it before the demo PR if possible because it removes C-string parloops. It also includes updates for Clement interpolation, which might be useful for students working on comparing methods for Hessian recovery? If you don't have time to look at this then I can self-review it.

**Comment by jwallwork23** [comment]
Note: has been rebased on top of the #139 branch so that will need to be merged first.

**Comment by jwallwork23** [comment]
Closes #141.

File: pyroteus/interpolation.py [hunk 1]
```diff
@@ -1,256 +1,263 @@
...
+    if not boundary:
+        dX = ufl.dx(domain=mesh)
+        volume = QualityMeasure(mesh, python=True)("volume")
+
+        # Compute patch volume
+        patch_volume = Function(P1)
+        domain = "{[i]: 0 <= i < patch.dofs}"
+        instructions = "patch[i] = patch[i] + vol[0]"
+        keys = {"vol": (volume, op2.READ), "patch": (patch_volume, op2.RW)}
+        firedrake.par_loop((domain, instructions), dX, keys)
```
Comment (stephankramer) [inline pyroteus/interpolation.py:82]: Yes, so these will need fixing (see #141)
I don't know if you want to merge this first and fix the par_loops later - or the other way around, and fix them on this branch?

Comment (jwallwork23) [reply pyroteus/interpolation.py:82]: Oh I see, I misunderstood your issue. I will self-review this and then get to that next.

File: pyroteus/recovery.py [hunk 2]
```diff
@@ -52,6 +32,39 @@ def recover_hessian(f: Function, method: str = "L2", **kwargs) -> Function:
...
+    """
+    msg = "Clement can only be used to compute gradients of"
+    if not isinstance(f, Function):
+        raise ValueError(f"{msg} Functions.")
+    family = f.ufl_element().family()
+    degree = f.ufl_element().degree()
+    if family not in ("Lagrange", "Discontinuous Lagrange"):
+        raise ValueError(f"{msg} Lagrange fields.")
+    if degree == 0:
+        raise ValueError(f"{msg} fields of degree > 0.")
```
Comment (jwallwork23) [inline pyroteus/recovery.py:51]: 'Functions' would be clearer than 'fields'.

File: pyroteus/recovery.py [hunk 3]
```diff
@@ -52,6 +32,39 @@ def recover_hessian(f: Function, method: str = "L2", **kwargs) -> Function:
...
+    mesh = f.function_space().mesh()
+
+    # Recover gradient
+    gradf = ufl.grad(f)
+    V = firedrake.VectorFunctionSpace(mesh, "DG", min(1, degree - 1))
+    if degree == 1:
+        g = clement_interpolant(firedrake.interpolate(gradf, V))
```
Comment (jwallwork23) [inline pyroteus/recovery.py:58]: Is this right? Surely Clement needs to be applied to a P0 field?

Comment (jwallwork23) [reply pyroteus/recovery.py]: See da4d3631fe0ccf910c8691affaabb90dae83205e.

**Pull request merged**